File: 2/Algo II/Practicos/pt6/TADNatural.c
typedef Nat = unsigned int;

fun cero () {
	unsigned int b = 0;
	return b;
}

fun sucesor (Nat x) {
	return x++:
}

fun mas (Nat x, Nat y){
	unsigned int r=0;
	if (x == 0) {
		r = y;
	}
	else if (y == 0) {
		r = x;
	}
	else {
		r = sucesor (mas(x-1,y-1));
	}
	return r;
}

fun por (Nat x, Nat y) {
	unsigned int r = 0;
	if (x==0 || y ==0) {
		r = 0;
	}
	else if (x == 1) {
		r = y;
	}
	else if (y == 1) {
		r = x;
	}
	else {
		r = por()
	}
}

File: 2/Algo II/Labs/TADS Frito/TADNatural.hs

module TADNatural (Natural,c,suc,mas,por,pot,men,menig,may,mayig,ig) where
    
    data Natural= Cero
                | Sucesor Natural

-- Operaciones
mas:: Natural -> Natural -> Natural
por, pot :: Natural -> Natural -> Natural
ig, may, mayig, men, menig, ig :: Natural -> Natural -> Bool

-- Ecuaciones

Cero `mas` x = x
Sucesor x `mas` y = Sucesor (x`mas`y)

Cero `por` x = Cero
Sucesor Cero `por` x = x
Sucesor x `por`y = Sucesor (x`por`y)

Cero `pot` Cero = "Indeterminado"
x `pot` Cero = Sucesor Cero
Sucesor x `pot` y = Sucesor (x`pot`y)

Cero `ig` Cero = True
Sucesor x `ig` Sucesor y = x `ig` y

Sucesor x `may` Cero = True
Sucesor x `may` y = Sucesor (x `may` y)

Cero `mayig` Cero = True
Sucesor x `mayig` Cero = True
Sucesor x `mayig` y | x `may` y = True    
                    | Sucesor x `ig` y = True
                    | otherwise = False

Sucesor x `men` Cero = False
Sucesor x `men` y = x `men` y

Cero `menig` Cero = True
Sucesor x `menig` Cero = False
Sucesor x `menig` y | x `men` y = True
                    | Sucesor x `ig` y = True
                    | otherwise = False
